app/config.go

```diff
@@ -507,6 +513,62 @@ var configurers = []configurer{
 		}
 		return nil
 	},
+
+	// ENABLE_GRPC
+	func(c *Config) error {
+		if experimental, _ := lookupBool("EXPERIMENTAL", false); experimental {
+			val, err := lookupBool("ENABLE_GRPC", false)
+			if err == nil {
+				c.EnableGRPC = val
+			}
+			return err
+		}
+		return nil
+	},
+
+	// TLS_CERT and TLS_KEY pair for gRPC server
```

Comment:
This is for gRPC [mutual-TLS authentication](https://grpc.io/docs/guides/auth.html#with-server-authentication-ssltls).

Reply:
What's gained from authenticating a client's cert? Is this intended to replace the `Origin` check on REST requests?

Reply:
Yes. It's also optional in our case, which I should have documented, so local development can be done with these values left empty. If both of these env vars are set, then the gRPC server will authenticate the clients with mTLS, otherwise its served `.WithInsecure()`. See:

https://github.com/keratin/authn-server/blob/f71d462bd316d2f599c467030df43b0436161cba/grpc/server/server.go#L41-L50

Reply:
Thanks for pointing that out, it does appear to make development easier.

Now my primary concerns are managing code/configuration complexity and understanding choices that I'll be responsible for maintaining. Please bear with me, I'm coming from a different operating environment and may not have the same context as you do.

I'm still uncertain what's gained from authenticating a gRPC client. The `Origin` check exists only to prevent XSRF attacks. Is that a concern with gRPC?

Or is this more of a replacement for HTTPS Basic Auth to secure communication between backends?

If AuthN didn't support mTLS, what would that mean for your deployment topology? Do you (would you) have a load balancer/proxy/sidecar that terminates TLS?

Reply:
A replacement for basic auth and TLS termination - if you happen to be deploying client certificates this means you can dispatch with TLS terminating sidecars altogether.

Reply:
Suppose some day that AuthN backend clients are updated to use gRPC and the admin REST endpoints are deprecated. Would that make client certs required for authenticating between backends?

Separately, suppose that AuthN is responsible for terminating TLS with a frontend client. Does that mean there's no load balancer, and if so, how could AuthN scale beyond a single instance?

Reply:
@Mohammed90 yeah that probably makes sense, in principle you could dispense with the per request authz via Basic Auth if you chose to make that implicit with the client cert, but you could have a situation where a service has access to the client cert but isn't mean to make admin requests to the admin interface. It's also more consistent.

> if the gRPC service is served insecured

I don't think anyone is suggesting it be unsecured. For public gRPC would need to TLS terminate the connection at a proxy/load balancer (just like we to with the public HTTP interface) - gRPC uses HTTP/2 so we can do this with a proxy that supports it. I think the main point of enabling TLS would be for service-to-service backend communication. Now there we could also use a service mesh thingy to terminate the TLS, but it seems like less hassle and more supported. The setup I would have would be mTLS on private backend connection and a load balancer terminating the gRPC connection for public connection (see for example https://blog.getambassador.io/securing-grpc-connections-with-tls-ssl-and-ambassador-bdd3ea0396c6).

Alternatively you could take the same approach as with HTTP as define it as out-of-scope for authn-server. Something like Istio can give you the mTLS on the backend and Envoy on the front end, so it would work. I'd be tempted to support direct TLS on both transports but I see why you haven't on HTTP. I'd say the difference for gRPC relates to the different practice around gRPC and maybe the availability of suitable h2c proxies, which you might not bother with in the backend.

Reply:
Let's clear things up a bit first, this is for [Mutual Authentication](https://en.m.wikipedia.org/wiki/Mutual_authentication), not for serving the service on HTTPS. Different layers.

Reply:
>Separately, suppose that AuthN is responsible for terminating TLS with a frontend client. Does that mean there's no load balancer, and if so, how could AuthN scale beyond a single instance

There is nothing to stop you from terminating TLS at multiple instances of authn all with the correct cert. Though it doesn't work for retries because the load balancer can't decrypt the connection and get the status code (e.g. 500).

Reply:
Okay, I'll eat my words. Even the load-balancer/reverse-proxy needs to authenticate with the certs if the gRPC server is configured as such, which makes sense in retrospect 😓 I guess I'm open to whatever decision you come up with for this.

Reply:
I'd like to consider this functionality in some future issue or pull request. When features can be separated, that's my preference. It sounds like the gRPC API can still achieve the same security as the REST API without it.

One advantage to approaching this later is that we might consider it more holistically. If AuthN begins to assume responsibility for connection security, should it also affect the REST API? Public, private, or both? Those questions will be easier to discuss when the answers don't feel like scope creep in an already meaningful changeset.

Reply:
Ok, I've removed the TLS configuration.